From 12669ce84c51576875d681ad6e280a55b7f3a606 Mon Sep 17 00:00:00 2001
From: Oliver Eftevaag <oliver.eftevaag@qt.io>
Date: Fri, 19 Sep 2025 14:52:53 +0200
Subject: [PATCH 090/239] QQuickStyle: Be smarter about detecting if a theme is
 dark

On Windows 11, the colorScheme is 'Unknown' when a contrast theme is
applied. This is because contrast themes are seperate to the classic
light/dark mode settings, but share the similarity that they both
determine palette colors. Some of our QQC2 themes provide their own
colors (like the Basic style, and FluentWinUI3 style). These styles
provide different colors depending on whether they think the system is
in dark mode or not. However, since the colorScheme can be 'Unknown',
and officially still have a dark-ish system palette, we need to be a
little more intelligent.

Also consider the system to be in "dark mode", in cases where the
colorScheme is 'Unknown' and the windowText is brighter than the window
palette color roles.

Task-number: QTBUG-140507
Change-Id: I179ce77104f0d682f353aec10d0f7f28660668b7
Reviewed-by: Doris Verria <doris.verria@qt.io>
(cherry picked from commit a28bb528ccebf5f5b999e1a16a1527b23a2e5445)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quickcontrols/qquickstyle.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/quickcontrols/qquickstyle.cpp b/src/quickcontrols/qquickstyle.cpp
index 7ac62244ce..566b06a356 100644
--- a/src/quickcontrols/qquickstyle.cpp
+++ b/src/quickcontrols/qquickstyle.cpp
@@ -388,8 +388,11 @@ const QPalette *QQuickStylePrivate::readPalette(const QSharedPointer<QSettings>
 bool QQuickStylePrivate::isDarkSystemTheme()
 {
     const bool dark = [](){
-        if (const QPlatformTheme *theme = QGuiApplicationPrivate::platformTheme())
+        if (const QPlatformTheme *theme = QGuiApplicationPrivate::platformTheme()) {
+            if (theme->colorScheme() == Qt::ColorScheme::Unknown)
+                return theme->palette()->windowText().color().lightnessF() > theme->palette()->window().color().lightnessF();
             return theme->colorScheme() == Qt::ColorScheme::Dark;
+        }
         return false;
     }();
     return dark;
-- 
2.51.2

